### PR TITLE
[MAJOR] Clean-up / Implementing subset of ml-agents package

### PR DESCRIPTION
**Pull request by alhasacademy96**
### Proposed change(s)

_The PR introduces significant changes to animal-ai-python:_

1. It revokes the dependency of ml-agents package as a whole and implements only ml-agents.envs sub package. 
In doing so, the package becomes substantially lighter and faster to download/install. Furthermore, the package reduces the chances of dependency issues the user may experience, such as in the past.

2. The package has new '`core tests`' folder, which must be passed before any new PR (this PR included). The core tests try to test out the functionality of the foundational code, and to make sure nothing breaks with the changes/additions/deletions brought by the PR. A series of tests are added but more will be added in the near-future. In addition, we now implement CI/CD when publishing new versions of animal-ai-python, such as on PyPI. The tests are also included in the workflow which are automated and fully embedded. 

3. With the help of the team (Wout, Kozzy), the repo is organized and made modern by utilizing on PEP 440 standards (1). 
4. Added a new script `update_dependencies`, to automate updating dependencies when required more efficiently (see script for comments). It is called only when attempting to publish to PyPI. If this fails, the error is caught and outputted (regarding dependency issues). 

5. Cleaned up README.md for a fresh look. 

6. Cleaned/renamed some scripts for readability and organization.

**A note to bear is that with the changes made especially on only using ml-agents.envs as opposed to ml-agents, this and all future versions of animal-ai-python are backwards incompatible**.

### Useful links (Github issues, ML-Agents forum threads etc.)

(1) - (https://sethmlarson.dev/pep-440)

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [ ] Code refactor
- [x] Breaking change
- [x] Documentation update
- [x] Repository Structure changes 
- [ ] Other (please describe)

### Checklist

- [x] Added tests that prove my fix is effective or that my feature works
- [x] Updated the changelog (if applicable)
- [x] Updated the documentation (if applicable)
- [ ] Updated the migration guide (if applicable)

### Other comments

All tests under `test/core` have passed my unit and functional testing for this PR. 

I've also tested the workflow CI/CD locally and everything works. (used hatch to build and staged publishing with twine).  

_Finally, This PR is still a WIP so more tests/documentation (where necessary) will be added._

### CODE REVIEW

File: test/integration/braitenberg.py
Comment (wschella) [inline test/integration/braitenberg.py:11]: Typo

**Comment by wschella** [comment]
This is quite hard to review, since there are no commit messages and (a lot of) formatting changes are mixed with behavior / code changes.

I won't block the merge or mingle too much, as I am still on sick leave.

One important comment tho: the update-dependencies script is invalid. All dependencies should be installed at the same time, otherwise pip will not guarantee that the subdependencies are conflict-free (i.e. they might be overridden by each install). In any case, do existing tools not handle this already, e.g. `pip install --upgrade .`?

**Comment by alhasacademy96** [comment]
There are commit descriptions as I like to keep the commit messages simple and short

**Comment by wschella** [comment]
Apologies, I now see they are in the message instead of the title.

**Comment by alhasacademy96** [comment]
The purpose of the update_dependencies.py script is to update the dependencies only during publishing to pypi, with additional safety checks. They are all updated at the same time if any of the allowed dependencies are not valid, hence the safety feature.

**Comment by alhasacademy96** [comment]
also please dont feel like you have to review as i know you are still recovering. i only included you as this is your branch and i felt like you might want to review just in case :)

**Comment by wschella** [comment]
> The purpose of the update_dependencies.py script is to update the dependencies only during publishing to pypi, with additional safety checks. They are all updated at the same time if any of the allowed dependencies are not valid, hence the safety feature.

I don't understand. Is this executed locally or in the CI pipeline?

If for CI: Our dependencies are not bundled in our pypi build, so the exact versions don't matter (version resolving will happen on the user end). What's the benefit over just installing from the pyproject.toml (which in CI will also pull the latest versions)?

If for locally: what's the difference as opposed to just editing pyproject.toml. and installing from that? Now it's two package versions to keep in sync.

In any case `pip install package1==v1 && pip install package2==v2 && ...` (what you are doing) is different than `pip install package1==v1 package2==v2 ...`. We need the latter to have consistent dependencies.

**Comment by alhasacademy96** [comment]
An excellent point you brought up. Here's my understanding and explanation:

**So, update_dependencies.py vs. pyproject.toml:**

_update_dependencies.py:_

This script can be used to enforce specific versions of dependencies, for example during testing or experimentation. It's useful when the CI pipeline or local development environment uses specific versions, rather than the latest ones, which we use specific ones. I just thought that it might be handly when using tests against specific versions of dependencies or even when we actually have compatibility issues regarding i.e. numpy. 

_pyproject.toml:_

For consistent dependency resolution, it's indeed better to install all dependencies in a single command like you suggested. This can be achieved by running pip install -r requirements.txt (which we're not implementing) or pip install (which installs the current package and its dependencies as specified in pyproject.toml (formerly setup.py for us). 

I do understand it's another layer of management on top of pyproject.toml but i just wanted to add the flexibility for more controlled testing environments as pyproject.toml uses the latest version of packages. Also, bear in mind that if there is nothing to be changed in update_dependencies.py (specifically allowed_versions), then the versions of dependencies will be identical to that of defined in pyproject.toml.

This can be problematic as we are using very specific versions of, say, ml-agents, numpy, protobuf, etc, and can't simply use their latest versions as it arises compatibility issues, and will probably continue to do so even if we migrate to ml-agents.env 1.0.0. 

However, this is the reason for reviews and collaboration and it doesnt mean we will adopt an extra layer of management automatically (which, i will be managing) so i didnt think it would affect the end user but rather us. It's for the flexibility that i proposed to using this extra layer. 

I hope i explained my reasoning for the changes.

P.S.: I am looking into implementing Docker and this change would certainly compliment it (although pyproject.toml is still a streamlined process regardless).

**Review by benaslater — CHANGES_REQUESTED** [review]

File: animalai/__init__.py
Comment (wschella) [reply animalai/__init__.py:2]: It allows setting up animalai without requiring the user to download extra files and set-up folders. Useful to keep examples and setup instructions short. Together with auto-downloading the binary in future versions, we could have a `pip install animalai` + 5 lines of demo code as setup instructions.

File: animalai/arenas/__init__.py
Comment (wschella) [reply animalai/arenas/__init__.py:1]: Cause they are the simplest possible arena's. Just to test everything works. We should maybe prefix with `example` to make this more clear.

File: README.md
Comment (alhasacademy96) [reply README.md:19]: The new link is the standard file that GitHub accepts for official contribution-related files - it's the same file but the location is different :)

File: animalai/__init__.py
Comment (alhasacademy96) [reply animalai/__init__.py:2]: We're still not implementing the auto-downloading feature so if it's meant to be just for that, then we don't need it right now. Can you confirm @wschella

File: animalai/arenas/__init__.py
Comment (alhasacademy96) [reply animalai/arenas/__init__.py:1]: I'm not sure why we have these two arenas to test things as they're too basic and do not really test anything but loading the arena with the agent and a single reward, with fixed and random positions.

File: test/core/README.md
Comment (alhasacademy96) [reply test/core/README.md:1]: I think either way works but my point was to emphasise the importance of these files to pass during unit/functional testing. Unit testing can be core aspects but also generic for specific features. what we want to do here is make sure these tests are passed as priority during development.

File: test/core/README.md
Comment (alhasacademy96) [reply test/core/README.md:9]: apologies, fixed now :)

Comment (alhasacademy96) [reply test/core/README.md:9]: I think it was my improper sentencing - fixed now :)

File: test/core/README.md
Comment (alhasacademy96) [reply test/core/README.md:26]: Fixed :)

File: animalai/__init__.py
Comment (wschella) [reply animalai/__init__.py:2]: No, it's not just for that. It's useful already to make initial setup (instructions) more simple. E.g. I already used the m in the integration tests for braitenberg and play mode.

File: animalai/arenas/__init__.py
Comment (wschella) [reply animalai/arenas/__init__.py:1]: Well, that tests a lot already right? As a user, this would be a very useful signal for me, in that I can successfully load non-empty arenas (and start playing / training agents). These very tasks are actually the tasks you would use to sanity check your agent implementation: Can it chase a simple green goal?

See also https://github.com/Kinds-of-Intelligence-CFI/animal-ai-python/pull/6#discussion_r1469731247

File: test/core/core-test-play.py
Comment (alhasacademy96) [reply test/core/core-test-play.py]: I'm not actually sure if we want to implement a manual test for running AAI for the reason that it's just challenging to due to the many interactive factors at play, such as defining the many parameters that can be configured. However, I've looked at mocking the environment to test that the main loop of the code does not produce any errors. This is basic and does not take into account the (advanced) configurable parameters of the arena but it ensures that the main loop of the code does not produce any errors upon launching. See the updated file.

Also, see here for further info on mock tests which I used to make one for manual play: https://docs.python.org/3/library/unittest.mock.html.

File: test/test-Configs/sanitryArenaCycling.yml
Comment (alhasacademy96) [reply test/test-Configs/sanitryArenaCycling.yml]: They're for anything to do with testing that Isnt core/foundational and whether if a new feature is added (as an example), these configs should be used for testing manually (addition to making sure the core tests pass before merge).

Comment (alhasacademy96) [reply test/test-Configs/sanitryArenaCycling.yml]: document updated

File: animalai/arenas/__init__.py
Comment (alhasacademy96) [reply animalai/arenas/__init__.py:1]: The logic for starting up AAI already contains code to handle instances where there are no arenas defined or the arenas are empty (upcoming in the next version of animalai). I think perhaps we could implement tests specifically with starting training mode, and with the example of an agent chasing a reward, this would be a better test to carry out.

Thoughts on this?

File: update_dependencies.py
Comment (alhasacademy96) [reply update_dependencies.py:1]: That is correct. I think feedback from wout also suggested that this file is redundant for now at least. I will wait for kozzy's thoughts on this for final decision @kozzy97

File: animalai/arenas/__init__.py
Comment (benaslater) [reply animalai/arenas/__init__.py:1]: I think it's good to distinguish between testing (for us to ensure AAI works) and sanity checking (for new users / setups to check things are working). I think these can be useful for both.
- If we want people to use these for sanity checks we should document that (e.g. the last step in the setup instructions is to run one of these)
- For testing I think a 'train to good reward on a really basic env' E2E test sounds good. An alternative that would mean we don't have to do training (since we don't need to test that) would be to use fixed action sequences. If we wanted to set a very high bar we could use these for screenshot testing (load into the env and the visual output has to look exactly like this)
I think someone mentioned renaming to example-arenas or something like that. I think that's a good idea since it makes it clear at a glance there's no functionality here

File: test/core/README.md
Comment (benaslater) [reply test/core/README.md:1]: If either way works for you I would bias towards the established terminology and not refer to these as core tests (I spent a few minutes googling core testing to try and figure out what was meant here).

I also don't think I agree with the idea that some tests are more critical than others - either we want AAI to do something or not. I think unless test duration is becoming prohibitive we might as well run all the tests that we have.

File: animalai/arenas/__init__.py
Comment (alhasacademy96) [reply animalai/arenas/__init__.py:1]: I see what you mean. I misunderstood it. Yep totally agree

File: pyproject.toml
Comment (alhasacademy96) [reply pyproject.toml:32]: A few factors are in play which dictates the specified version of ml-agents we are using, mainly that ml-agents is deprecated and new versions are rare (there is a newer version but we should'nt expect massive changes or improvements since funding for the project is withheld indefinitely).

Python versioning is strict for 0.30.0 and for 1.0.0 (new version of ml-agents).  Also, there were a few dependency issues, namely numpy, protobuf, which required to be specified as listed on ml-agents documentation. However, these can change if/when we upgrade to ml-agents 1.0.0 after evaluation and testing.

File: pyproject.toml
Comment (alhasacademy96) [reply pyproject.toml:34]: See above comment

File: animalai/actions.py
Comment (alhasacademy96) [reply animalai/actions.py:6]: I agree that having the random method is a redundancy, which is simply a wrapper as i understand it (used for np.random.choice()). Can we call this function directly on the class's attributes? 
Also, if random is not doing anything significant, can't we remove it to reduec overhead?  However, i do think that the random method can make things easier to read and understand which is why i believe @wschella used it (correct me if im wrong).

Comment (alhasacademy96) [reply animalai/actions.py:6]: I've never used dataclasses so i think @wschella is the person to decide here.

Comment (alhasacademy96) [reply animalai/actions.py:6]: Is AAIActions.random() easier to understand/use than np.random.choice(AAIActions.allActions)? Especially for those who aren't familiar with NumPy or can't instantly spot allActions.

File: animalai/actions.py
Comment (alhasacademy96) [reply animalai/actions.py:90]: see above comment

File: animalai/actions.py
Comment (wschella) [reply animalai/actions.py:6]: I don't think I implemented this. This was already here, if anything, I just formatted / made sure typing was correct.

I wouldn't necessarily touch it. An enum would definitely make sense, but from what I know this is also quite standard pythonic coding for this sort of static collection.

File: test/core/core-test-play.py
Comment (wschella) [reply test/core/core-test-play.py]: I'm a fan to keep some manual test scripts in the repo, when I'm making changes I tend to fire one of these when a automated (unit)test fails, as I can freely make some changes to debug more closely and then discard later. It also helps to assert that it's the actual code failing, and it's not a problem in the mocking/unittest (which tends to happen sometime for succifiently complex mocking).

Comment (wschella) [reply test/core/core-test-play.py]: In any case, it shouldn't be in the unittest /automated test folder (whether we keep it or not).

**Pull request merged**
